[processor/logstransformprocessor]: Race condition on shutdown

### Component(s)

processor/logstransform

### What happened?

## Description

In the logstransform processor, it looks like the pipeline is shutdown, then the converter, then the fromConverter, which triggers the shutdown of the various goroutines here.

It's possible the pipeline and converter are quickly shutdown, while `converterLoop` is in the middle of processing entries:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ff1325cab96dd16527ccd2b8e073d1e7639658bd/processor/logstransformprocessor/processor.go#L153-L159

Since we don't guarantee that converter loop is stopped before shutting down the pipeline, it's possible for entries to be sent down the shut-down pipeline, where it can end up panicking the LogEmitter by sending on a closed channel (e.flush here sends on a channel that is closed when Close is called).
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ff1325cab96dd16527ccd2b8e073d1e7639658bd/pkg/stanza/adapter/emitter.go#L114-L120

## Steps to Reproduce
Shutting down the collector while a high-throughput pipeline is using the logstransform processor.

## Expected Result
No panic

## Actual Result
```
panic: send on closed channel

goroutine 101 [running]:
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/adapter.(*LogEmitter).flush(...)
	/home/runner/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.92.0/adapter/emitter.go:159
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/adapter.(*LogEmitter).Process(0xc0007f25a0, {0x87b4960, 0xc000a8ab40}, 0x0?)
	/home/runner/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.92.0/adapter/emitter.go:116 +0xb8
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper.(*WriterOperator).Write(0xc00030a700, {0x87b4960, 0xc000a8ab40}, 0xc00462c480?)
	/home/runner/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.92.0/operator/helper/writer.go:53 +0xd6
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper.(*ParserOperator).ProcessWithCallback(0xc00030a700, {0x87b4960, 0xc000a8ab40}, 0x10?, 0x1?, 0x0)
	/home/runner/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.92.0/operator/helper/parser.go:122 +0xa5
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper.(*ParserOperator).ProcessWith(...)
	/home/runner/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.92.0/operator/helper/parser.go:98
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/parser/csv.(*Parser).Process(0xc000a71f00, {0x87b4960, 0xc000a8ab40}, 0x0)
	/home/runner/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.92.0/operator/parser/csv/csv.go:140 +0x176
github.com/open-telemetry/opentelemetry-collector-contrib/processor/logstransformprocessor.(*logsTransformProcessor).converterLoop(0xc0008d9800, {0x87b4960, 0xc000a8ab40})
	/home/runner/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/logstransformprocessor@v0.92.0/processor.go:155 +0x349
created by github.com/open-telemetry/opentelemetry-collector-contrib/processor/logstransformprocessor.(*logsTransformProcessor).Start
	/home/runner/go/pkg/mod/github.com/open-telemetry/opentelemetry-collector-contrib/processor/logstransformprocessor@v0.92.0/processor.go:115 +0x470
```


### Collector version

v0.92.0

### Environment information

_No response_

### OpenTelemetry Collector configuration

_No response_

### Log output

_No response_

### Additional context

_No response_

--- a/.chloggen/fix_log-transform-shutdown.yaml
+++ b/.chloggen/fix_log-transform-shutdown.yaml
@@ -1,0 +1,13 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: logstransformprocessor
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Fix potential panic on shutdown due to incorrect shutdown order
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [31139]

--- a/processor/logstransformprocessor/processor.go
+++ b/processor/logstransformprocessor/processor.go
@@ -78,8 +78,14 @@ func (ltp *logsTransformProcessor) Shutdown(ctx context.Context) error {
 }
 
 func (ltp *logsTransformProcessor) Start(ctx context.Context, _ component.Host) error {
+	// create all objects before starting them, since the consumer loops depend on the converters not being nil.
+	ltp.converter = adapter.NewConverter(ltp.logger)
+
+	wkrCount := int(math.Max(1, float64(runtime.NumCPU())))
+	ltp.fromConverter = adapter.NewFromPdataConverter(wkrCount, ltp.logger)
+
 	// data flows like this:
-	// fromConverter -> converter loop -> pipeline -> adapter -> emitter loop -> converter -> consumer loop
+	// fromConverter -> converter loop -> pipeline -> emitter loop -> converter -> consumer loop
 	// We should start the pipeline in reverse order, and stop it in this order.
 	ltp.startConsumerLoop(ctx)
 	ltp.startConverter()
@@ -95,9 +101,6 @@ func (ltp *logsTransformProcessor) Start(ctx context.Context, _ component.Host) 
 }
 
 func (ltp *logsTransformProcessor) startFromConverter() {
-	wkrCount := int(math.Max(1, float64(runtime.NumCPU())))
-
-	ltp.fromConverter = adapter.NewFromPdataConverter(wkrCount, ltp.logger)
 	ltp.fromConverter.Start()
 
 	ltp.shutdownFns = append(ltp.shutdownFns, func(ctx context.Context) error {
@@ -111,7 +114,7 @@ func (ltp *logsTransformProcessor) startFromConverter() {
 func (ltp *logsTransformProcessor) startConverterLoop(ctx context.Context) {
 	wg := &sync.WaitGroup{}
 	wg.Add(1)
-	go ltp.converterLoop(wg, ctx)
+	go ltp.converterLoop(ctx, wg)
 
 	ltp.shutdownFns = append(ltp.shutdownFns, func(ctx context.Context) error {
 		wg.Wait()
@@ -144,7 +147,8 @@ func (ltp *logsTransformProcessor) startPipeline() error {
 func (ltp *logsTransformProcessor) startEmitterLoop(ctx context.Context) {
 	wg := &sync.WaitGroup{}
 	wg.Add(1)
-	go ltp.emitterLoop(wg, ctx)
+	go ltp.emitterLoop(ctx, wg)
+
 	ltp.shutdownFns = append(ltp.shutdownFns, func(ctx context.Context) error {
 		wg.Wait()
 		return nil
@@ -152,21 +156,21 @@ func (ltp *logsTransformProcessor) startEmitterLoop(ctx context.Context) {
 }
 
 func (ltp *logsTransformProcessor) startConverter() {
-	ltp.converter = adapter.NewConverter(ltp.logger)
 	ltp.converter.Start()
+
 	ltp.shutdownFns = append(ltp.shutdownFns, func(ctx context.Context) error {
 		ltp.converter.Stop()
 		return nil
 	})
 }
 
+// startConsumerLoop starts the loop which reads all the logs produced by the converter
+// (aggregated by Resource) and then places them on the next consumer
 func (ltp *logsTransformProcessor) startConsumerLoop(ctx context.Context) {
-	// ...
-	// * third which reads all the logs produced by the converter
-	//   (aggregated by Resource) and then places them on the next consumer
 	wg := &sync.WaitGroup{}
 	wg.Add(1)
-	go ltp.consumerLoop(wg, ctx)
+	go ltp.consumerLoop(ctx, wg)
+
 	ltp.shutdownFns = append(ltp.shutdownFns, func(ctx context.Context) error {
 		wg.Wait()
 		return nil
@@ -180,7 +184,7 @@ func (ltp *logsTransformProcessor) ConsumeLogs(_ context.Context, ld plog.Logs) 
 
 // converterLoop reads the log entries produced by the fromConverter and sends them
 // into the pipeline
-func (ltp *logsTransformProcessor) converterLoop(wg *sync.WaitGroup, ctx context.Context) {
+func (ltp *logsTransformProcessor) converterLoop(ctx context.Context, wg *sync.WaitGroup) {
 	defer wg.Done()
 
 	for {
@@ -208,7 +212,7 @@ func (ltp *logsTransformProcessor) converterLoop(wg *sync.WaitGroup, ctx context
 
 // emitterLoop reads the log entries produced by the emitter and batches them
 // in converter.
-func (ltp *logsTransformProcessor) emitterLoop(wg *sync.WaitGroup, ctx context.Context) {
+func (ltp *logsTransformProcessor) emitterLoop(ctx context.Context, wg *sync.WaitGroup) {
 	defer wg.Done()
 
 	for {
@@ -230,7 +234,7 @@ func (ltp *logsTransformProcessor) emitterLoop(wg *sync.WaitGroup, ctx context.C
 }
 
 // consumerLoop reads converter log entries and calls the consumer to consumer them.
-func (ltp *logsTransformProcessor) consumerLoop(wg *sync.WaitGroup, ctx context.Context) {
+func (ltp *logsTransformProcessor) consumerLoop(ctx context.Context, wg *sync.WaitGroup) {
 	defer wg.Done()
 
 	for {

--- a/processor/logstransformprocessor/processor.go
+++ b/processor/logstransformprocessor/processor.go
@@ -114,7 +114,8 @@ func (ltp *logsTransformProcessor) startConverterLoop(ctx context.Context) {
 	go ltp.converterLoop(wg, ctx)
 
 	ltp.shutdownFns = append(ltp.shutdownFns, func(ctx context.Context) error {
-		return waitForWgOrCtx(wg, ctx)
+		wg.Wait()
+		return nil
 	})
 }
 
@@ -145,7 +146,8 @@ func (ltp *logsTransformProcessor) startEmitterLoop(ctx context.Context) {
 	wg.Add(1)
 	go ltp.emitterLoop(wg, ctx)
 	ltp.shutdownFns = append(ltp.shutdownFns, func(ctx context.Context) error {
-		return waitForWgOrCtx(wg, ctx)
+		wg.Wait()
+		return nil
 	})
 }
 
@@ -166,7 +168,8 @@ func (ltp *logsTransformProcessor) startConsumerLoop(ctx context.Context) {
 	wg.Add(1)
 	go ltp.consumerLoop(wg, ctx)
 	ltp.shutdownFns = append(ltp.shutdownFns, func(ctx context.Context) error {
-		return waitForWgOrCtx(wg, ctx)
+		wg.Wait()
+		return nil
 	})
 }
 
@@ -247,20 +250,4 @@ func (ltp *logsTransformProcessor) consumerLoop(wg *sync.WaitGroup, ctx context.
 			}
 		}
 	}
-}
-
-func waitForWgOrCtx(wg *sync.WaitGroup, ctx context.Context) error {
-	doneChan := make(chan struct{})
-	go func() {
-		wg.Wait()
-		close(doneChan)
-	}()
-
-	select {
-	case <-ctx.Done():
-		return ctx.Err()
-	case <-doneChan: // OK
-	}
-
-	return nil
 }

--- a/processor/logstransformprocessor/processor.go
+++ b/processor/logstransformprocessor/processor.go
@@ -32,7 +32,6 @@ type logsTransformProcessor struct {
 	emitter       *adapter.LogEmitter
 	converter     *adapter.Converter
 	fromConverter *adapter.FromPdataConverter
-	wg            sync.WaitGroup
 	shutdownFns   []component.ShutdownFunc
 }
 
@@ -64,25 +63,69 @@ func (ltp *logsTransformProcessor) Capabilities() consumer.Capabilities {
 }
 
 func (ltp *logsTransformProcessor) Shutdown(ctx context.Context) error {
-	for _, fn := range ltp.shutdownFns {
+	ltp.logger.Info("Stopping logs transform processor")
+	// We call the shutdown functions in reverse order, so that the last thing we started
+	// is stopped first.
+	for i := len(ltp.shutdownFns) - 1; i >= 0; i-- {
+		fn := ltp.shutdownFns[i]
+
 		if err := fn(ctx); err != nil {
 			return err
 		}
 	}
-	ltp.wg.Wait()
 
 	return nil
 }
 
 func (ltp *logsTransformProcessor) Start(ctx context.Context, _ component.Host) error {
+	// data flows like this:
+	// fromConverter -> converter loop -> pipeline -> adapter -> emitter loop -> converter -> consumer loop
+	// We should start the pipeline in reverse order, and stop it in this order.
+	ltp.startConsumerLoop(ctx)
+	ltp.startConverter()
+	ltp.startEmitterLoop(ctx)
+	err := ltp.startPipeline()
+	if err != nil {
+		return err
+	}
+	ltp.startConverterLoop(ctx)
+	ltp.startFromConverter()
 
+	return nil
+}
+
+func (ltp *logsTransformProcessor) startFromConverter() {
+	wkrCount := int(math.Max(1, float64(runtime.NumCPU())))
+
+	ltp.fromConverter = adapter.NewFromPdataConverter(wkrCount, ltp.logger)
+	ltp.fromConverter.Start()
+
+	ltp.shutdownFns = append(ltp.shutdownFns, func(ctx context.Context) error {
+		ltp.fromConverter.Stop()
+		return nil
+	})
+}
+
+// startConverterLoop starts the converter loop, which reads all the logs translated by the fromConverter and then forwards
+// them to pipeline
+func (ltp *logsTransformProcessor) startConverterLoop(ctx context.Context) {
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go ltp.converterLoop(wg, ctx)
+
+	ltp.shutdownFns = append(ltp.shutdownFns, func(ctx context.Context) error {
+		return waitForWgOrCtx(wg, ctx)
+	})
+}
+
+func (ltp *logsTransformProcessor) startPipeline() error {
 	// There is no need for this processor to use storage
 	err := ltp.pipe.Start(storage.NewNopClient())
 	if err != nil {
 		return err
 	}
+
 	ltp.shutdownFns = append(ltp.shutdownFns, func(ctx context.Context) error {
-		ltp.logger.Info("Stopping logs transform processor")
 		return ltp.pipe.Stop()
 	})
 
@@ -92,40 +135,39 @@ func (ltp *logsTransformProcessor) Start(ctx context.Context, _ component.Host) 
 	}
 	ltp.firstOperator = pipelineOperators[0]
 
-	wkrCount := int(math.Max(1, float64(runtime.NumCPU())))
+	return nil
+}
 
+// startEmitterLoop starts the loop which reads all the logs modified by the pipeline and then forwards
+// them to converter
+func (ltp *logsTransformProcessor) startEmitterLoop(ctx context.Context) {
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go ltp.emitterLoop(wg, ctx)
+	ltp.shutdownFns = append(ltp.shutdownFns, func(ctx context.Context) error {
+		return waitForWgOrCtx(wg, ctx)
+	})
+}
+
+func (ltp *logsTransformProcessor) startConverter() {
 	ltp.converter = adapter.NewConverter(ltp.logger)
 	ltp.converter.Start()
 	ltp.shutdownFns = append(ltp.shutdownFns, func(ctx context.Context) error {
 		ltp.converter.Stop()
 		return nil
 	})
+}
 
-	ltp.fromConverter = adapter.NewFromPdataConverter(wkrCount, ltp.logger)
-	ltp.fromConverter.Start()
-	ltp.shutdownFns = append(ltp.shutdownFns, func(ctx context.Context) error {
-		ltp.fromConverter.Stop()
-		return nil
-	})
-	// Below we're starting 3 loops:
-	// * first which reads all the logs translated by the fromConverter and then forwards
-	//   them to pipeline
-	// ...
-	ltp.wg.Add(1)
-	go ltp.converterLoop(ctx)
-
-	// * second which reads all the logs modified by the pipeline and then forwards
-	//   them to converter
-	// ...
-	ltp.wg.Add(1)
-	go ltp.emitterLoop(ctx)
-
+func (ltp *logsTransformProcessor) startConsumerLoop(ctx context.Context) {
 	// ...
 	// * third which reads all the logs produced by the converter
 	//   (aggregated by Resource) and then places them on the next consumer
-	ltp.wg.Add(1)
-	go ltp.consumerLoop(ctx)
-	return nil
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go ltp.consumerLoop(wg, ctx)
+	ltp.shutdownFns = append(ltp.shutdownFns, func(ctx context.Context) error {
+		return waitForWgOrCtx(wg, ctx)
+	})
 }
 
 func (ltp *logsTransformProcessor) ConsumeLogs(_ context.Context, ld plog.Logs) error {
@@ -135,8 +177,8 @@ func (ltp *logsTransformProcessor) ConsumeLogs(_ context.Context, ld plog.Logs) 
 
 // converterLoop reads the log entries produced by the fromConverter and sends them
 // into the pipeline
-func (ltp *logsTransformProcessor) converterLoop(ctx context.Context) {
-	defer ltp.wg.Done()
+func (ltp *logsTransformProcessor) converterLoop(wg *sync.WaitGroup, ctx context.Context) {
+	defer wg.Done()
 
 	for {
 		select {
@@ -163,8 +205,8 @@ func (ltp *logsTransformProcessor) converterLoop(ctx context.Context) {
 
 // emitterLoop reads the log entries produced by the emitter and batches them
 // in converter.
-func (ltp *logsTransformProcessor) emitterLoop(ctx context.Context) {
-	defer ltp.wg.Done()
+func (ltp *logsTransformProcessor) emitterLoop(wg *sync.WaitGroup, ctx context.Context) {
+	defer wg.Done()
 
 	for {
 		select {
@@ -185,8 +227,8 @@ func (ltp *logsTransformProcessor) emitterLoop(ctx context.Context) {
 }
 
 // consumerLoop reads converter log entries and calls the consumer to consumer them.
-func (ltp *logsTransformProcessor) consumerLoop(ctx context.Context) {
-	defer ltp.wg.Done()
+func (ltp *logsTransformProcessor) consumerLoop(wg *sync.WaitGroup, ctx context.Context) {
+	defer wg.Done()
 
 	for {
 		select {
@@ -205,4 +247,20 @@ func (ltp *logsTransformProcessor) consumerLoop(ctx context.Context) {
 			}
 		}
 	}
+}
+
+func waitForWgOrCtx(wg *sync.WaitGroup, ctx context.Context) error {
+	doneChan := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(doneChan)
+	}()
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-doneChan: // OK
+	}
+
+	return nil
 }

--- a/processor/logstransformprocessor/processor.go
+++ b/processor/logstransformprocessor/processor.go
@@ -84,9 +84,17 @@ func (ltp *logsTransformProcessor) Start(ctx context.Context, _ component.Host) 
 	wkrCount := int(math.Max(1, float64(runtime.NumCPU())))
 	ltp.fromConverter = adapter.NewFromPdataConverter(wkrCount, ltp.logger)
 
-	// data flows like this:
-	// fromConverter -> converter loop -> pipeline -> emitter loop -> converter -> consumer loop
-	// We should start the pipeline in reverse order, and stop it in this order.
+	// data flows in this order:
+	// ConsumeLogs: receives logs and forwards them for conversion to stanza format ->
+	// fromConverter: converts logs to stanza format ->
+	// converterLoop: forwards converted logs to the stanza pipeline ->
+	// pipeline: performs user configured operations on the logs ->
+	// emitterLoop: forwards output stanza logs for conversion to OTLP ->
+	// converter: converts stanza logs to OTLP ->
+	// consumerLoop: sends the converted OTLP logs to the next consumer
+	//
+	// We should start these components in reverse order of the data flow, then stop them in order of the data flow,
+	// in order to allow for pipeline draining.
 	ltp.startConsumerLoop(ctx)
 	ltp.startConverter()
 	ltp.startEmitterLoop(ctx)

--- a/processor/logstransformprocessor/processor.go
+++ b/processor/logstransformprocessor/processor.go
@@ -78,7 +78,7 @@ func (ltp *logsTransformProcessor) Shutdown(ctx context.Context) error {
 }
 
 func (ltp *logsTransformProcessor) Start(ctx context.Context, _ component.Host) error {
-	// create all objects before starting them, since the consumer loops depend on the converters not being nil.
+	// create all objects before starting them, since the loops (consumerLoop, converterLoop) depend on these converters not being nil.
 	ltp.converter = adapter.NewConverter(ltp.logger)
 
 	wkrCount := int(math.Max(1, float64(runtime.NumCPU())))

--- a/processor/logstransformprocessor/processor_test.go
+++ b/processor/logstransformprocessor/processor_test.go
@@ -198,7 +198,7 @@ func generateLogData(messages []testLogMessage) plog.Logs {
 }
 
 // laggy operator is a test operator that simulates heavy processing that takes a large amount of time.
-// The heavey processing only occurs for every 100th log
+// The heavy processing only occurs for every 100th log
 type laggyOperator struct {
 	helper.WriterOperator
 	logsCount int
@@ -280,9 +280,4 @@ func TestProcessorShutdownWithSlowOperator(t *testing.T) {
 
 	err = ltp.Shutdown(context.Background())
 	require.NoError(t, err)
-
-	// We should ensure that all channels are drained and all logs
-	// are put into ConsumeLogs are emitted to the consumer at this point,
-	// but this is not the case right now.
-	// require.Equal(t, 500, tln.LogRecordCount())
 }

--- a/processor/logstransformprocessor/processor_test.go
+++ b/processor/logstransformprocessor/processor_test.go
@@ -5,7 +5,6 @@ package logstransformprocessor
 
 import (
 	"context"
-	"sync"
 	"testing"
 	"time"
 
@@ -195,23 +194,4 @@ func generateLogData(messages []testLogMessage) plog.Logs {
 	}
 
 	return ld
-}
-
-func TestWaitForWgOrCtx(t *testing.T) {
-	t.Run("Waitgroup is done, context is not", func(t *testing.T) {
-		wg := &sync.WaitGroup{}
-		err := waitForWgOrCtx(wg, context.Background())
-		require.NoError(t, err)
-	})
-
-	t.Run("Context is done, wait group is not", func(t *testing.T) {
-		wg := &sync.WaitGroup{}
-		wg.Add(1)
-
-		ctx, cancel := context.WithCancel(context.Background())
-		cancel()
-
-		err := waitForWgOrCtx(wg, ctx)
-		require.ErrorIs(t, err, context.Canceled)
-	})
 }

--- a/processor/logstransformprocessor/processor_test.go
+++ b/processor/logstransformprocessor/processor_test.go
@@ -206,12 +206,9 @@ type laggyOperator struct {
 
 func (t *laggyOperator) Process(ctx context.Context, e *entry.Entry) error {
 
+	// Wait for a large amount of time every 100 logs
 	if t.logsCount%100 == 0 {
-		select {
-		case <-time.After(100 * time.Millisecond): // OK
-			// case <-ctx.Done(): // operator should return immediately
-			// 	return nil
-		}
+		time.Sleep(100 * time.Millisecond)
 	}
 
 	t.logsCount++

--- a/processor/logstransformprocessor/processor_test.go
+++ b/processor/logstransformprocessor/processor_test.go
@@ -5,6 +5,7 @@ package logstransformprocessor
 
 import (
 	"context"
+	"sync"
 	"testing"
 	"time"
 
@@ -194,4 +195,23 @@ func generateLogData(messages []testLogMessage) plog.Logs {
 	}
 
 	return ld
+}
+
+func TestWaitForWgOrCtx(t *testing.T) {
+	t.Run("Waitgroup is done, context is not", func(t *testing.T) {
+		wg := &sync.WaitGroup{}
+		err := waitForWgOrCtx(wg, context.Background())
+		require.NoError(t, err)
+	})
+
+	t.Run("Context is done, wait group is not", func(t *testing.T) {
+		wg := &sync.WaitGroup{}
+		wg.Add(1)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		cancel()
+
+		err := waitForWgOrCtx(wg, ctx)
+		require.ErrorIs(t, err, context.Canceled)
+	})
 }

--- a/processor/logstransformprocessor/processor_test.go
+++ b/processor/logstransformprocessor/processor_test.go
@@ -14,6 +14,7 @@ import (
 	"go.opentelemetry.io/collector/pdata/pcommon"
 	"go.opentelemetry.io/collector/pdata/plog"
 	"go.opentelemetry.io/collector/processor/processortest"
+	"go.uber.org/zap"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/testdata"
 	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatatest/plogtest"
@@ -194,4 +195,97 @@ func generateLogData(messages []testLogMessage) plog.Logs {
 	}
 
 	return ld
+}
+
+// laggy operator is a test operator that simulates heavy processing that takes a large amount of time.
+// The heavey processing only occurs for every 100th log
+type laggyOperator struct {
+	helper.WriterOperator
+	logsCount int
+}
+
+func (t *laggyOperator) Process(ctx context.Context, e *entry.Entry) error {
+
+	if t.logsCount%100 == 0 {
+		select {
+		case <-time.After(100 * time.Millisecond): // OK
+			// case <-ctx.Done(): // operator should return immediately
+			// 	return nil
+		}
+	}
+
+	t.logsCount++
+
+	t.Write(ctx, e)
+	return nil
+}
+
+func (t *laggyOperator) CanProcess() bool {
+	return true
+}
+
+type laggyOperatorConfig struct {
+	helper.WriterConfig
+}
+
+func (l *laggyOperatorConfig) Build(s *zap.SugaredLogger) (operator.Operator, error) {
+	wo, err := l.WriterConfig.Build(s)
+	if err != nil {
+		return nil, err
+	}
+
+	return &laggyOperator{
+		WriterOperator: wo,
+	}, nil
+}
+
+func TestProcessorShutdownWithSlowOperator(t *testing.T) {
+	operator.Register("laggy", func() operator.Builder { return &laggyOperatorConfig{} })
+
+	config := &Config{
+		BaseConfig: adapter.BaseConfig{
+			Operators: []operator.Config{
+				{
+					Builder: func() *laggyOperatorConfig {
+						l := &laggyOperatorConfig{}
+						l.OperatorType = "laggy"
+						return l
+					}(),
+				},
+			},
+		},
+	}
+
+	tln := new(consumertest.LogsSink)
+	factory := NewFactory()
+	ltp, err := factory.CreateLogsProcessor(context.Background(), processortest.NewNopCreateSettings(), config, tln)
+	require.NoError(t, err)
+	assert.True(t, ltp.Capabilities().MutatesData)
+
+	err = ltp.Start(context.Background(), nil)
+	require.NoError(t, err)
+
+	testLog := plog.NewLogs()
+	scopeLogs := testLog.ResourceLogs().AppendEmpty().
+		ScopeLogs().AppendEmpty()
+
+	for i := 0; i < 500; i++ {
+		lr := scopeLogs.LogRecords().AppendEmpty()
+		lr.Body().SetStr("Test message")
+	}
+
+	// The idea is to check that shutdown, when there are a lot of entries, doesn't try to write logs to
+	// a closed channel, since that'll cause a panic.
+	// In order to test, we send a lot of logs to be consumed, then shutdown immediately.
+
+	err = ltp.ConsumeLogs(context.Background(), testLog)
+	require.NoError(t, err)
+
+	err = ltp.Shutdown(context.Background())
+	require.NoError(t, err)
+
+	// We should ensure that all channels are drained and all logs
+	// are put into ConsumeLogs are emitted to the consumer at this point,
+	// but this is not the case right now.
+	// require.Equal(t, 500, tln.LogRecordCount())
 }

